### Introduction
---

Boolean  ```bool```

Integer  ```int```

Rational (```int/int```) ```fractions.Fraction```

Real     ```float``` or ```deciaml.Decimal```

Complex  ```complex```

### Integer: ```int```
---

**Data type**

In [1]:
import sys

sys.getsizeof(0)  # 24 bytes overhead

24

In [2]:
sys.getsizeof(1)  # 4 bytes to store 1

28

In [3]:
sys.getsizeof(2**1000)  # 4 bytes to store 1

160

In [4]:
def calc(a):
  for i in range(10000000):
    a * 2

In [5]:
import time 


for n in [1, 100, 10000]:
  start = time.perf_counter()
  a = 2**n
  calc(a)
  end = time.perf_counter()
  print(f"{n}:  {round(end-start,5)} seconds")

1:  0.34827 seconds
100:  0.58334 seconds
10000:  4.17296 seconds


**Operation**

```python
n = (n // d) * d + (n % d)
```

```floor(a) = b``` is ```int(b) <= a```, e.g. ```floor(3.14) = 3```, ```float(-3.14) = -4```

```//``` is ```floor()``` of the division, e.g. ```135 // 4 = 33```, ```-135 // 4 = -34```

**Exercise**

In [6]:
def from_base10(n: int, b: int):
  if b < 2:
    raise ValueError('Base b must be 2 or more (>= 2)')
  if n < 0:
    raise ValueError('Number n must be positive')
  digits = []
  while n > 0:
    n, m = divmod(n, b)
    digits.insert(0, m)
  return digits

In [7]:
from_base10(10, 2)

[1, 0, 1, 0]

In [8]:
from_base10(255, 16)

[15, 15]

In [9]:
def encode(digits: list, m: str):
  if max(digits) >= len(m):
    raise ValueError('Map m is not long enough to encode the digits')
    
  return ''.join([m[d] for d in digits])

```python
''.join([m[d] for d in digits])
```
is better than

```python
encoding = ''
  for d in digits:
    encoding += m[d]
```

as ```encoding += m[d]``` is creating a new string each time


In [10]:
import string

def rebase_from10(number, base):
  
  # Create encoding map
  digits_map = ''.join([str(n) for n in range(10)]) + string.ascii_uppercase
  if base < 2 or base > 36:
    raise ValueError('Base b must be between 2 and 36')
    
  # Allow for negative number
  sign = -1 if number < 0 else 1
  number *= sign
  
  digits = from_base10(number, base)
  encoded = encode(digits, digits_map)
  
  if sign == -1:
    encoded = '-' + encoded

  return encoded

In [11]:
e = rebase_from10(10, 2)
print(e)
print(int(e, base=2))

1010
10


In [12]:
e = rebase_from10(-23865, 16)
print(e)
print(int(e, base=16))

-5D39
-23865


### Rational number: ```fractions.Fraction```
---

```float``` objects have **finite** precision $\implies$ **any** ```float``` can be written as a fraction!

In [13]:
import math
from fractions import Fraction

x = Fraction(math.pi)
x

Fraction(884279719003555, 281474976710656)

In [14]:
x = x.limit_denominator(100)
x

Fraction(311, 99)

In [15]:
Fraction(float(1/8))

Fraction(1, 8)

In [16]:
Fraction(float(3/10))

Fraction(5404319552844595, 18014398509481984)

### Floats: ```float```
---

Python ```float``` = C ```double```

```float``` uses ```64 bits```:

> sign $\Rightarrow$ 1 bit

> exponent $\Rightarrow$ 11 bits $\Rightarrow$ range [-1022, 1023]

> significant $\Rightarrow$ 52 bits $\Rightarrow$ 15-17 significant (base-10) digits

**Internal representation**

In [17]:
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __int__(self, /)
 |      int(sel

In [18]:
float(Fraction(22,7))

3.142857142857143

In [19]:
format(0.1,'0.25f')

'0.1000000000000000055511151'

In [20]:
format(0.125,'0.25f')

'0.1250000000000000000000000'

In [21]:
a = 0.1 + 0.1 + 0.1
b = 0.3
a == b

False

In [22]:
format(a,'0.25f')

'0.3000000000000000444089210'

In [23]:
format(b,'0.25f')

'0.2999999999999999888977698'

**Equality testing**

1. Round both sides, ```round(a, 5) == round(b, 5)```, however ```round()``` can cause inaccuracies

2. Use an appropriate tolerance (**PEP 485**), ``` a = b if and only if |a - b| < tol ```:

> Relative tolerances, ```rel_tol```, for large numbers

> Absolute tolerance, ```abs_tol```, for numbers near zeros 

In [24]:
x = 1000.0000001
y = 1000.0000002
math.isclose(x, y)

True

In [25]:
help(math.isclose)

Help on built-in function isclose in module math:

isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
    Determine whether two floating point numbers are close in value.
    
      rel_tol
        maximum difference for being considered "close", relative to the
        magnitude of the input values
      abs_tol
        maximum difference for being considered "close", regardless of the
        magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [26]:
# rel_tol=1e-09, abs_tol=0.0
x = 0.0000001
y = 0.0000002
math.isclose(x, y)

False

In [27]:
# Setting abs_tol
x = 0.0000001
y = 0.0000002
math.isclose(x, y, abs_tol=1e-5)

True

**Coercing float to integer**

1. ```math.trunc()```

2. ```math.floor()``` (```math.trunc()``` for positive numbers) and ```math.ceil()``` (```math.trunc()``` for negative numbers)

3. ```round()```

In [28]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



> **Normal rounding**: round to the nearest value, with ties round **away from zero**

> **Banker's rounding**: IEEE 754 standard - rounds to the nearest value, with ties round to the nearest value with an **even least significant digit**

Banker's rounding is standard by ```round()``` as it is *less biased* rounding than ties away from zero

In [29]:
# Tie between 1.3 and 1.4, rounds away from zero
round(1.35, 1)

1.4

In [30]:
# Tie between 1.2 and 1.3, rounds towards zero
round(1.25, 1)

1.2

In [31]:
round(15.00, -1)

20.0

In [32]:
round(25.00, -1)

20.0

### Decimals: ```decimal``` (PEP 327)
---

Exact representation for base 10 decimals

Particularly important for finance, when summing *many* transactions

**Construct**

Decimals have a **context** that controls certain aspects, for example:

> ```prec``` (mathematical operation, not construction)

> ```rounding```

In [46]:
import decimal
from decimal import Decimal

print(f"Global context: {decimal.getcontext()}")
print()

print(f"Banker's rounding: {decimal.getcontext().rounding}")
print()

print(f"Local context manager: {decimal.localcontext(ctx=None)}")
print()

Global context: Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

Banker's rounding: ROUND_HALF_EVEN

Local context manager: <decimal.ContextManager object at 0x0000018B7A36E130>



In [48]:
# Within the 'with' bubble
with decimal.localcontext() as ctx:
  ctx.prec = 6
  ctx.rounding = decimal.ROUND_HALF_UP
  print(decimal.getcontext())
  print()
  
print(decimal.getcontext())

Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])


In [52]:
format(1.2, '.20f')

'1.19999999999999995559'

In [61]:
# str
a = Decimal('1.2')           # Yes

# float
b = Decimal(1.2)             # NO! It is storing 1.19999999999999995559

# tuple (sign, s.d., expo)
c = Decimal((0, (1,2), -1))  # Yes = -1^[0] * [12] * 10^-[1]

In [62]:
a == b

False

In [63]:
a == c

True

In [70]:
a = Decimal('0.123456') 
b = Decimal('0.123457') 
print(a+b)

with decimal.localcontext() as ctx:
  ctx.prec = 2
  ctx.rounding = decimal.ROUND_HALF_UP
  c = a+b
  print(a, b, c)
  
print(c)

0.246913
0.123456 0.123457 0.25
0.25


**Math operations**

```python
n = (n // d) * d + (n % d)
```

In [82]:
n = 10
d = 3

print(divmod(n, d))
print(divmod(Decimal(str(n)), Decimal(str(d))))

(3, 1)
(Decimal('3'), Decimal('1'))


In [83]:
n = -10
d = 3

print(divmod(n, d))
print(divmod(Decimal(str(n)), Decimal(str(d))))

(-4, 2)
(Decimal('-3'), Decimal('-1'))


In [85]:
print(-4*3 + 2)
print(-3*3 + -1)

-10
-10


In [92]:
a = Decimal('0.01')
print(format(a.sqrt() * a.sqrt(), '1.27f'))           # Yes
print(format(math.sqrt(a) * math.sqrt(a), '1.27f'))   # NO - Casting Decimal to float before sqrt

0.010000000000000000000000000
0.010000000000000001942890293


**Performance consideration**

> Not easy to code (construction via string and truples, not numerical literals)

> Not all mathematical functions in ```math``` exist in ```decimal```

> More memory overhead

> Performance is much slower than floats (relatively)

In [95]:
# Memory footprint

a = 3.1415
b = Decimal(str(a))

In [96]:
sys.getsizeof(a)

24

In [97]:
sys.getsizeof(b)

104

In [109]:
# Speed performance

import time 
n = 10000000

In [110]:
# Construction

def run_float(n=1):
  for i in range(n):
    a = 3.1415
    
def run_decimal(n=1):
  for i in range(n):
    a = Decimal('3.1415')

start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print(f"float:   {round(end-start, 5)} s")

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print(f"decimal: {round(end-start, 5)} s")

float:   0.18064 s
decimal: 3.11674 s


In [111]:
# Adding

def add_float(n=1):
  a = 3.1415
  for i in range(n):
    a + a
    
def add_decimal(n=1):
  a = Decimal('3.1415')
  for i in range(n):
    a + a

start = time.perf_counter()
add_float(n)
end = time.perf_counter()
print(f"float:   {round(end-start, 5)} s")

start = time.perf_counter()
add_decimal(n)
end = time.perf_counter()
print(f"decimal: {round(end-start, 5)} s")

float:   0.33881 s
decimal: 0.70714 s


In [112]:
# Square root
n = 5000000

def sqrt_float(n=1):
  a = 3.1415
  for i in range(n):
    math.sqrt(a)
    
def sqrt_decimal(n=1):
  a = Decimal('3.1415')
  for i in range(n):
    a.sqrt()

start = time.perf_counter()
sqrt_float(n)
end = time.perf_counter()
print(f"float:   {round(end-start, 5)} s")

start = time.perf_counter()
sqrt_decimal(n)
end = time.perf_counter()
print(f"decimal: {round(end-start, 5)} s")

float:   0.58237 s
decimal: 12.63149 s


### Complex: ```complex```
---

In [122]:
d = 2 - 3j

In [123]:
print(type(d))

<class 'complex'>


In [124]:
print(d.real)
print(d.imag)
print(d.conjugate())

2.0
-3.0
(2+3j)


In [125]:
print(d == -3j + 2)

True


In [126]:
import cmath

print(cmath.phase(d))

-0.982793723247329


Euler's identity

$e^{i\pi} + 1 = 0$.

In [136]:
LHS = 0
RHS = cmath.exp(cmath.pi * 1j) + 1
RHS

1.2246467991473532e-16j

In [137]:
cmath.isclose(RHS, LHS)

False

In [138]:
cmath.isclose(RHS, LHS, abs_tol=1e-5)

True

### Boolean: ```bool``` (PEP 285)
---

```bool``` is **subclass** of ```int```

```bool``` posses all properties of ```int``` and has some specialized ones, e.g. ```and```, ```or```

```bool```'s objects ```True``` and ```False``` are ```singleton``` objects, therefore they always retain their same memory address throughout the lifetime.

**Note**: ```True == 1``` BUT ```True is not 1``` as ```id(True) != id(1)```

In [145]:
print(issubclass(bool, int))
print(issubclass(int, bool))

True
False


In [157]:
print(True is 1)
print(True == 1)

False
True


In [158]:
# Strange expression

print(True > False)
print()

print((1==2) is False)
print((1==2) == False)
print((1==2) == 0)
print((1==2) is 0)

True

True
True
True
False


In [159]:
print(True + True)

2
